File: sdks/python/apache_beam/io/pubsub.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""Google Cloud PubSub sources and sinks.

Cloud Pub/Sub sources and sinks are currently supported only in streaming
pipelines, during remote execution.
"""

from __future__ import absolute_import

from apache_beam import coders
from apache_beam.io import iobase


class PubSubSource(iobase.NativeSource):
  """Source for reading from a given Cloud Pub/Sub topic.

  Attributes:
    topic: Cloud Pub/Sub topic in the form "/topics/<project>/<topic>".
    subscription: Optional existing Cloud Pub/Sub subscription to use in the
      form "projects/<project>/subscriptions/<subscription>".
    id_label: The attribute on incoming Pub/Sub messages to use as a unique
      record identifier.  When specified, the value of this attribute (which can
      be any string that uniquely identifies the record) will be used for
      deduplication of messages.  If not provided, Dataflow cannot guarantee
      that no duplicate data will be delivered on the Pub/Sub stream. In this
      case, deduplication of the stream will be strictly best effort.
    coder: The Coder to use for decoding incoming Pub/Sub messages.
  """

  def __init__(self, topic, subscription=None, id_label=None,
               coder=coders.StrUtf8Coder()):
    self.topic = topic
    self.subscription = subscription
    self.id_label = id_label
    self.coder = coder

  @property
  def format(self):
    """Source format name required for remote execution."""
    return 'pubsub'

  def reader(self):
    raise NotImplementedError(
        'PubSubSource is not supported in local execution.')


class PubSubSink(iobase.NativeSink):
  """Sink for writing to a given Cloud Pub/Sub topic."""

  def __init__(self, topic, coder=coders.StrUtf8Coder()):
    self.topic = topic
    self.coder = coder

  @property
  def format(self):
    """Sink format name required for remote execution."""
    return 'pubsub'

  def writer(self):
    raise NotImplementedError(
        'PubSubSink is not supported in local execution.')

File: sdks/typescript/src/apache_beam/io/pubsub.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as PubSub from "@google-cloud/pubsub";

import * as beam from "../pvalue";
import * as external from "../transforms/external";
import * as internal from "../transforms/internal";
import { AsyncPTransform, withName } from "../transforms/transform";
import { RowCoder } from "../coders/row_coder";
import { BytesCoder } from "../coders/required_coders";
import { requireForSerialization } from "../serialization";
import { packageName } from "../utils/packageJson";
import { serviceProviderFromJavaGradleTarget } from "../utils/service";
import { camelToSnakeOptions } from "../utils/utils";

const PUBSUB_EXPANSION_GRADLE_TARGET =
  "sdks:java:io:google-cloud-platform:expansion-service:shadowJar";

const readSchema = RowCoder.inferSchemaOfJSON({
  topic: "string",
  subscription: "string",
  idAttribute: "string",
  timestampAttribute: "string",
});

type ReadOptions =
  | {
      topic: string;
      subscription?: never;
      idAttribute?: string;
      timestampAttribute?: string;
    }
  | {
      topic?: never;
      subscription: string;
      idAttribute?: string;
      timestampAttribute?: string;
    };

// TODO: Schema-producing variants.
export function readFromPubSub(
  options: ReadOptions,
): AsyncPTransform<beam.Root, beam.PCollection<Uint8Array>> {
  if (options.topic && options.subscription) {
    throw new TypeError(
      "Exactly one of topic or subscription must be provided.",
    );
  }
  return withName(
    "readFromPubSubRaw",
    external.rawExternalTransform<beam.Root, beam.PCollection<Uint8Array>>(
      "beam:transform:org.apache.beam:pubsub_read:v1",
      camelToSnakeOptions(options),
      serviceProviderFromJavaGradleTarget(PUBSUB_EXPANSION_GRADLE_TARGET),
    ),
  );
}

function readFromPubSubWithAttributesRaw(
  options: ReadOptions,
): AsyncPTransform<beam.Root, beam.PCollection<Uint8Array>> {
  if (options.topic && options.subscription) {
    throw new TypeError(
      "Exactly one of topic or subscription must be provided.",
    );
  }
  return withName(
    "readFromPubSubWithAttributesRaw",
    external.rawExternalTransform<beam.Root, beam.PCollection<Uint8Array>>(
      "beam:transform:org.apache.beam:pubsub_read:v1",
      { needsAttributes: true, ...camelToSnakeOptions(options) },
      serviceProviderFromJavaGradleTarget(PUBSUB_EXPANSION_GRADLE_TARGET),
    ),
  );
}

export function readFromPubSubWithAttributes(
  options: ReadOptions,
): AsyncPTransform<
  beam.Root,
  beam.PCollection<PubSub.protos.google.pubsub.v1.PubsubMessage>
> {
  return async function readFromPubSubWithAttributes(root: beam.Root) {
    return (
      await root.applyAsync(readFromPubSubWithAttributesRaw(options))
    ).map((encoded) =>
      PubSub.protos.google.pubsub.v1.PubsubMessage.decode(encoded),
    );
  };
}

type WriteOptions = { idAttribute?: string; timestampAttribute?: string };

function writeToPubSubRaw(
  topic: string,
  options: WriteOptions = {},
): AsyncPTransform<beam.PCollection<Uint8Array>, {}> {
  return withName(
    "writeToPubSubRaw",
    external.rawExternalTransform<beam.PCollection<Uint8Array>, {}>(
      "beam:transform:org.apache.beam:pubsub_write:v1",
      { topic, ...camelToSnakeOptions(options) },
      serviceProviderFromJavaGradleTarget(PUBSUB_EXPANSION_GRADLE_TARGET),
    ),
  );
}

export function writeToPubSub(topic: string, options: WriteOptions = {}) {
  return async function writeToPubSub(dataPColl: beam.PCollection<Uint8Array>) {
    return dataPColl //
      .map((data) =>
        PubSub.protos.google.pubsub.v1.PubsubMessage.encode({ data }).finish(),
      )
      .apply(internal.withCoderInternal(new BytesCoder()))
      .applyAsync(writeToPubSubRaw(topic, options));
  };
}

requireForSerialization(`${packageName}/io/pubsub`, PubSub);
